File: api/src/main/java/org/taymyr/openapi/OpenAPIService.java
package org.taymyr.openapi;

import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.Descriptor;
import com.lightbend.lagom.javadsl.api.Service;
import com.lightbend.lagom.javadsl.api.ServiceCall;

import static com.lightbend.lagom.javadsl.api.Service.pathCall;

import static java.lang.String.format;

/**
 * OpenAPI service descriptor.
 *
 * @author Sergey Morgunov
 */
public interface OpenAPIService extends Service {

    /**
     * Return OpenAPI specification for current service.
     * @return OpenAPI specification
     */
    ServiceCall<NotUsed, String> openapi();

    default Descriptor withOpenAPI(Descriptor descriptor) {
        return descriptor.withCalls(
                pathCall(format("/%s/openapi", descriptor.name()), this::openapi)
        );
    }

}

File: java/api/src/main/kotlin/org/taymyr/lagom/javadsl/openapi/OpenAPIService.kt
@file:JvmName("OpenAPIUtils")
@file:Suppress("DeprecatedCallableAddReplaceWith")
package org.taymyr.lagom.javadsl.openapi

import akka.NotUsed
import com.lightbend.lagom.javadsl.api.Descriptor
import com.lightbend.lagom.javadsl.api.Service
import com.lightbend.lagom.javadsl.api.Service.pathCall
import com.lightbend.lagom.javadsl.api.ServiceCall
import java.util.Optional
import kotlin.reflect.jvm.javaMethod

/**
 * OpenAPI service descriptor.
 */
@Deprecated("Use `OpenAPIRouter` instead")
interface OpenAPIService : Service {

    /**
     * Return OpenAPI specification for current service.
     * @param format Format of OpenAPI specification. Can be `json` or `yaml`.
     * @return OpenAPI specification
     */
    fun openapi(format: Optional<String>): ServiceCall<NotUsed, String>

    @JvmDefault
    @Deprecated("Use static function OpenAPIUtils.withOpenAPI(...) instead.")
    fun withOpenAPI(descriptor: Descriptor): Descriptor = descriptor.withOpenAPI()
}

@Deprecated("Use `OpenAPIRouter` instead")
fun Descriptor.withOpenAPI(): Descriptor = this.withCalls(
    pathCall<NotUsed, String>("/_${this.name()}/openapi?format", OpenAPIService::openapi.javaMethod)
)
